### PR TITLE
Create a tutorial for using moveit_opw_kinematics_plugin

### PR DESCRIPTION
**Pull request by simonschmeisser**
### Description

This adds a tutorial for the moveit_opw_kinematics_plugin written by @JeroenDM based on opw_kinematics by Jeremy Meyer

### CODE REVIEW

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+The `opw_kinematics_plugin https://github.com/JeroenDM/moveit_opw_kinematics_plugin`_ can be installed from packages: ::
+
+  sudo apt install ros-noetic-opw-kinematics-plugin
+
+Usage
+------
+
+- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.
+- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``
+- Set parameters to describe the geometry of your manipulator. The following is an example for Kuka Kr6R700: ::
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:33]:
```suggestion
- Set parameters to describe the geometry of your manipulator. The following is an example for the KUKA KR6 R700: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      a2: -0.035
+      b:   0.000
+      c1:  0.400
+      c2:  0.315
+      c3:  0.365
+      c4:  0.080
+    opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
+    opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]
+
+The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will throw ROS_ERRORs on startup if they
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:48]:
```suggestion
The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print ``ROS_ERROR``s on startup if they
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+do not match your URDF, so you can safely guess and test if needed:
+
+.. image:: images/opw.png
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-I robot description packages. This has already started for Fanuc.
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:56]:
```suggestion
We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
```

not sure this is really specific to support packages though.

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:57]:
```suggestion
Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+an example you can reduce your *kinematics.yaml* to the following header: ::
+
+  manipulator:
+    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
+
+and then add second line of the following to your *launch/planning_context.launch*: ::
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:63]:
```suggestion
and then add a ``rosparam`` ``load`` line to your *launch/planning_context.launch* which causes the parameters in that file to be loaded onto the parameter server: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
+  <group ns="$(arg robot_description)_kinematics">
+    <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
+    <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
+  </group>
+
+Note that the *ns* parameter has to match the name you gave your planning group during the setup.
+
+This can also be done directly within the MoveIt setup assistant. On the *Define Planning Group* page, select the *opw_parameters*
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:73]:
```suggestion
The MoveIt Setup Assistant can automatically insert that line, removing the need to manually edit the ``planning_context.launch`` file. On the *Define Planning Group* page, select the *opw_parameters*
```

**Review by gavanderhoorn — COMMENTED** [review]
Some comments.

File: doc/opw_kinematics/opw_kinematics_tutorial.rst (hunk 1 above)
Comment (simonschmeisser) [reply doc/opw_kinematics/opw_kinematics_tutorial.rst:33]: just googled, it should actually be KR 6 R700
```suggestion
- Set parameters to describe the geometry of your manipulator. The following is an example for the KUKA KR 6 R700: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst (hunk 3 above)
Comment (simonschmeisser) [reply doc/opw_kinematics/opw_kinematics_tutorial.rst:56]: I'm not sure I understand your what you are not sure about. The intention of this paragraph was to include somewhere that collaboration on parameter files is desired

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      a2: -0.035
+      b:   0.000
+      c1:  0.400
+      c2:  0.315
+      c3:  0.365
+      c4:  0.080
+    opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
+    opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]
+
+The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print ``ROS_ERROR``s on startup if they
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:48]: Not sure the `s` works there after the `ROS_ERROR`.

At least GH highlighter seems to get confused.

Comment (simonschmeisser) [reply doc/opw_kinematics/opw_kinematics_tutorial.rst]: that seems easily confused, but http://rst.ninjs.org/ found some more issues, let's see if that also fixes travis

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+This package is meant to provide a simpler alternative to IK-Fast based solutions in situations where one has an
+industrial robot with a parallel base and spherical wrist. This configuration is extremely common in industrial robots.
+
+The kinematics are parameterized by 7 primary values taken directly from the robot's spec sheet and a set of
+joint-zero offsets. Given this structure, no other setup is required.
+
+Installation
+-------------
+
+The `opw_kinematics_plugin <https://github.com/JeroenDM/moveit_opw_kinematics_plugin>`_ can be installed from packages: ::
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:24]:
```suggestion
The `opw_kinematics_plugin <https://github.com/JeroenDM/moveit_opw_kinematics_plugin>`_ can be installed using ``apt`` on Ubuntu and Debian: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+  sudo apt install ros-noetic-opw-kinematics-plugin
+
+Usage
+------
+
+- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.
+- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:32]:
```suggestion
- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``.
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+- Set parameters to describe the geometry of your manipulator.
+
+The following is an example for the KUKA KR 6 R700: ::
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:35]:
```suggestion
The following is an example of the parameters needed for the KUKA KR 6 R700: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    opw_kinematics_geometric_parameters:
+      a1:  0.025
+      a2: -0.035
+      b:   0.000
+      c1:  0.400
+      c2:  0.315
+      c3:  0.365
+      c4:  0.080
+    opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
+    opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:48]:
```suggestion
.. code-block:: yaml

   manipulator:
     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
     opw_kinematics_geometric_parameters:
       a1:  0.025
       a2: -0.035
       b:   0.000
       c1:  0.400
       c2:  0.315
       c3:  0.365
       c4:  0.080
     opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
     opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]
```

And I believe you don't need the `::` any more with this.

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+do not match your URDF, so you can safely guess and test if needed:
+
+.. image:: images/opw.png
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
+Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
+an example you can reduce your *kinematics.yaml* to the following header: ::
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:60]:
```suggestion
an example you can reduce your ``kinematics.yaml`` to the following: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+  manipulator:
+    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
+
+and then add a ``rosparam`` ``load`` line to your *launch/planning_context.launch* which causes the parameters in that file to be loaded onto the parameter server: ::
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:65]:
```suggestion
and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server: ::
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+  <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
+  <group ns="$(arg robot_description)_kinematics">
+    <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
+    <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
+  </group>
+
+Note that the *ns* parameter has to match the name you gave your planning group during the setup.
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:73]:
```suggestion
Note that the ``ns`` parameter has to match the name you gave your planning group during the setup.
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+The MoveIt Setup Assistant can automatically insert that line, removing the need to manually edit the ``planning_context.launch`` file. On the *Define Planning Group* page, select the *opw_parameters*
+file for your robot in the *Kin. parameters file* field
+
+.. image:: images/assistant.png
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:78]: Would be great to highlight the input field in the screenshot (with a req square or something).

**Review by gavanderhoorn — COMMENTED** [review]
Some more comments.

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  sudo apt install ros-noetic-opw-kinematics-plugin
+
+Usage
+------
+
+- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.
+- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``.
+- Set parameters to describe the geometry of your manipulator.
+
+The following is an example of the parameters needed for the KUKA KR 6 R700: ::
```
Comment (simonschmeisser) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:35]:
```suggestion
The following is an example of the parameters needed for the KUKA KR 6 R700:
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
+Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
+an example you can reduce your ``kinematics.yaml`` to the following: ::
+
+  manipulator:
+    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:65]:
```suggestion
.. code-block:: yaml

   manipulator:
     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server: ::
+
+  <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
+  <group ns="$(arg robot_description)_kinematics">
+    <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
+    <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
+  </group>
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:73]:
```suggestion
.. code-block:: xml

   <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
   <group ns="$(arg robot_description)_kinematics">
     <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
     <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
   </group>
```

**Review by gavanderhoorn — COMMENTED** [review]
Apologies. Somehow I missed these.

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+do not match your URDF, so you can safely guess and test if needed:
+
+.. image:: images/opw.png
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
+Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
+an example you can reduce your ``kinematics.yaml`` to the following: ::
```
Comment (simonschmeisser) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:62]:
```suggestion
an example you can reduce your ``kinematics.yaml`` to the following:
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+  manipulator:
+    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
+
+and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server: ::
```
Comment (simonschmeisser) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:69]:
```suggestion
and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server:
```

**Review by gavanderhoorn — APPROVED** [review]
I believe this is OK like this now.

Only comment left: the tutorial links readers to the parameters for the LR Mate 200iB, but the screenshots shows an M-10iA and the config file in the input field is also for the M-10iA.

**Comment by simonschmeisser** [comment]
thanks for the throughout review, I hacked this together yesterday evening/night to have something contributed at WMD2021 and would/should have maybe left it as draft for some longer. Will change it to M-10iA (that screenshot is reused from the PR)

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print a ``ROS_ERROR`` on startup if they
+do not match your URDF, so you can safely guess and test if needed:
+
+.. image:: images/opw.png
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
+Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
```
Comment (simonschmeisser) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:61]:
```suggestion
Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```

Comment (simonschmeisser) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst]:
```suggestion
Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print a ``ROS_ERROR`` on startup if they
+do not match your URDF, so you can safely guess and test if needed:
+
+.. image:: images/opw.png
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
+Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```
Comment (simonschmeisser) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:61]:
```suggestion
Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```

File: doc/opw_kinematics/opw_kinematics_tutorial.rst [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print a ``ROS_ERROR`` on startup if they
+do not match your URDF, so you can safely guess and test if needed:
+
+.. image:: images/opw.png
+    
+Sharing OPW descriptions
+-------------------------
+
+We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
+Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```
Comment (gavanderhoorn) [inline doc/opw_kinematics/opw_kinematics_tutorial.rst:61]:
```suggestion
Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```
I would probably still recommend to use a permalink.

Comment (simonschmeisser) [reply doc/opw_kinematics/opw_kinematics_tutorial.rst:61]: Oh, I got confused why your previous suggestion wasn't applied any longer but now I see that your original suggestion was from branch -> hash already and I reverted that with my last change. Sorry!

**Review by JafarAbdi — APPROVED** [review]

**Pull request merged**